Enabling Unified Shared Memory (USM) interfaces

mmeterel opened the pull request:
This PR is adding Unified Shared Memory interfaces, functional tests and documentation for BLAS domain.

All unit tests are passing both locally and through CI.

USM interfaces are part of latest oneAPI spec (0.7)

mmeterel commented:
/intelci: run

petercad commented on cmake/FindMKL.cmake:
```diff
@@ -66,6 +66,8 @@ if(UNIX)
   set(LIB_PREFIX "-l")
   set(OPENCL_LIBNAME "OpenCL")
+  set(SYCL_LIB_PREFIX "-f")
+  set(SYCL_LIBNAME "sycl")
```
The `-fsycl` flag isn't specifying a library, but instead telling the linker that there are SYCL source files. So I think it would be more accurate to replace these two lines with e.g.

```
  set(SYCL_LINK_FLAGS "-fsycl")
```

and use `SYCL_LINK_FLAGS` below.

mmeterel replied:
Done

petercad commented on docs/domains/blas/blas-level-1-routines.rst:
```diff
@@ -0,0 +1,76 @@
+         * -     \ `swap <swap.html>`__\   
+           -     Vector-vector swap      
+         * -     \ `iamax <iamax.html>`__\   
+           -     Index of the maximum absolute value element of a       vector     
```
There's extra space here (between "a" and "vector") and in other places in the routine list files. But if it doesn't show up in the formatted output, feel free to ignore this comment :)

mmeterel replied:
I think, these came from the translator. I have been removing the ones I notice. I can remove this one as well :)

petercad commented on docs/domains/blas/blas-level-3-routines.rst:
```diff
@@ -0,0 +1,79 @@
+   -  
+
+
+      .. container::
```
Can we remove lines 47-63, which are CPU-specific?

mmeterel replied:
Done

petercad commented on docs/domains/blas/blas-like-extensions.rst:
```diff
@@ -0,0 +1,48 @@
+         * -     Routines
+           -     Description     
+         * -     \ `gemm_batch <gemm_batch.html>`__\   
+           -     Computes groups of matrix-matrix product with general       matrices.   
```
`product` -> `products`

mmeterel replied:
Done

petercad commented on include/onemkl/blas/blas.hpp:
```diff
@@ -1814,6 +1680,1928 @@ static inline void trsv(cl::sycl::queue &queue, uplo upper_lower, transpose tran
+
+static inline cl::sycl::event asum(cl::sycl::queue &queue, std::int64_t n,
+                                   const std::complex<float> *x, std::int64_t incx, float *result,
+                                   const cl::sycl::vector_class<cl::sycl::event> &dependencies) {
```
`dependencies` should default to an empty vector -- `dependencies = {}`

mmeterel replied:
Done

mmeterel commented on include/onemkl/blas/detail/blas_ct_templates.hpp:
```diff
@@ -0,0 +1,2109 @@
+static inline cl::sycl::event syr2(cl::sycl::queue &queue, uplo upper_lower, std::int64_t n,
+                                   float alpha, const float *x, std::int64_t incx, const float *y,
+                                   std::int64_t incy, float *a, std::int64_t lda,
+                                   const cl::sycl::vector_class<cl::sycl::event> &dependencies);
```
For this PR, I manually created all files. I updated the header file.

mmeterel commented on src/blas/backends/cublas/cublas_batch.cpp:
```diff
@@ -102,83 +59,123 @@ void gemm_batch(cl::sycl::queue &queue, transpose transa, transpose transb, std:
+                           int64_t group_count, int64_t *groupsize,
+                           const cl::sycl::vector_class<cl::sycl::event> &dependencies) {
+    throw backend_unsupported_exception();
+    return cl::sycl::event();
```
Done.

mmeterel commented on src/blas/backends/cublas/cublas_helper.hpp:
```diff
@@ -28,6 +28,12 @@
 #include <stdexcept>
 #include "onemkl/types.hpp"
+
+class backend_unsupported_exception : public std::runtime_error {
```
Created a new header file inside src/include.  I think, the goal is to keep src/ and tests/ independent.

mmeterel commented on src/blas/backends/mklgpu/mkl_blas_sycl_usm.cpp:
```diff
@@ -0,0 +1,1298 @@
+                     std::int64_t lda, const float *b, std::int64_t ldb, float beta, float *c,
+                     std::int64_t ldc,
+                     const cl::sycl::vector_class<cl::sycl::event> &dependencies) {
+    return onemkl::mklgpu::internal::gemm(queue, transa, transb, m, n, k, alpha, a, lda, b, ldb,
```
My understanding is, internal namespace is used to convert the enum parameters from onemkl to mkl. I followed the existing (buffer based API) structure in this PR.

mmeterel commented on src/blas/backends/mklgpu/mkl_internal_blas_gpu_wrappers.cpp:
```diff
@@ -1466,33 +1190,1337 @@ void gemm_ext(cl::sycl::queue &queue, onemkl::transpose transa, onemkl::transpos
 
+// USM APIs
+
+cl::sycl::event gemm(cl::sycl::queue &queue, onemkl::transpose transa, onemkl::transpose transb,
```
Done.

mmeterel commented on tests/unit_tests/blas/extensions/gemmt_usm.cpp:
```diff
@@ -0,0 +1,292 @@
+    for (int i = 0; i < C.size(); i++)
+        C_ref[i] = C[i];
+
+    // Call Reference GEMM.
```
Done.

mmeterel commented on src/blas/backends/mklgpu/mkl_internal_blas_gpu_wrappers.cpp:
```diff
@@ -1466,33 +1190,1337 @@ void gemm_ext(cl::sycl::queue &queue, onemkl::transpose transa, onemkl::transpos
+        coalesced_events.push_back(gemm_event);
+        total_groupsize += groupsize[i];
+    }
+#ifdef _WIN64
```
Done

mmeterel commented on src/blas/backends/mklgpu/mkl_internal_blas_gpu_wrappers.cpp:
```diff
@@ -1466,33 +1190,1337 @@ void gemm_ext(cl::sycl::queue &queue, onemkl::transpose transa, onemkl::transpos
+    }
+#ifdef _WIN64
+    for (int64_t j = 0; j < group_count; j++)
+        (*coalesced_events[j]).wait();
```
Done

mmeterel commented on src/blas/backends/mklgpu/mkl_internal_blas_sycl_gpu.hpp:
```diff
@@ -88,7 +88,7 @@ inline CBLAS_OFFSET cblas_convert(onemkl::offset o) {
 
```
Good point. I don't think, these are necessary. I removed them.

mmeterel commented on tests/unit_tests/blas/level1/axpy_usm.cpp:
```diff
@@ -0,0 +1,149 @@
+using namespace cl::sycl;
+using std::vector;
+
+extern std::vector<cl::sycl::device> devices;
```
Yes, it is passed to GTEST instantiation call.

mmeterel commented on tests/unit_tests/blas/level1/asum_usm.cpp:
```diff
@@ -0,0 +1,145 @@
+    // Compare the results of reference implementation and DPC++ implementation.
+
+    bool good;
+    { good = check_equal(*result_p, result_ref, N, std::cout); }
```
Fixed all applicable cases.

jasukhar commented:
/intelci: rerun

mmeterel commented:
/intelci: rerun

mmeterel commented on tests/unit_tests/blas/include/usm_allocator_helper.hpp:
```diff
@@ -0,0 +1,64 @@
+#include "test_helper.hpp"
+
+template <typename T, int align>
+class usm_allocator_helper : public cl::sycl::usm_allocator<T, cl::sycl::usm::alloc::shared, 64> {
```
Updated the code to use sycl::usm_allocator directly.

mmeterel commented on tests/unit_tests/blas/level1/asum_usm.cpp:
```diff
@@ -0,0 +1,145 @@
+    bool good;
+    { good = check_equal(*result_p, result_ref, N, std::cout); }
+
+    if (!dev.is_host())
```
Removed the if condition for all cases.

mmeterel commented on tests/unit_tests/blas/level1/axpy_usm.cpp:
```diff
@@ -0,0 +1,149 @@
+
+    y_ref.resize(y.size());
+    for (int i = 0; i < y.size(); i++)
+        y_ref[i] = y[i];
```
Code updated.

mmeterel commented on src/blas/backends/mklcpu/cpu_batch.cpp:
```diff
@@ -1104,5 +367,308 @@ void trsm_batch(cl::sycl::queue &queue, side left_right, uplo upper_lower, trans
+        host_task<class mkl_kernel_sgemm_batch_usm>(cgh, [=]() {
+            float **a_array = (float **)::malloc(sizeof(float *) * batch_size);
+            float **b_array = (float **)::malloc(sizeof(float *) * batch_size);
+            float **c_array = (float **)::malloc(sizeof(float *) * batch_size);
```
Decided to add nullptr checks since some of the memory allocations are using MKL specific types.

mmeterel commented:
Hi @petercad 

Thanks a lot for your valuable feedback. I believe I addressed all your comments. Please let me know if you have any further feedback.

mmeterel commented:
/intelci: restart

mmeterel commented:
/intelci: restart

mmeterel commented:
/intelci: run

mmeterel commented:
/intelci: run

mmeterel commented:
/intelci: run

mmeterel commented:
/intelci: restart

mmeterel commented:
/intelci: run

mmeterel commented:
/intelci: restart

mkrainiuk commented on cmake/FindMKL.cmake:
```diff
@@ -84,7 +85,7 @@ if (ENABLE_MKLCPU_BACKEND OR ENABLE_MKLGPU_BACKEND)
   endif()
   if(ENABLE_MKLGPU_BACKEND)
-    set(MKL_LINK_SYCL ${MKL_LINK_PREFIX} ${LIB_PREFIX}${MKL_SYCL} ${MKL_LINK_C} ${LIB_PREFIX}${OPENCL_LIBNAME})
+    set(MKL_LINK_SYCL ${MKL_LINK_PREFIX} ${LIB_PREFIX}${MKL_SYCL} ${MKL_LINK_C} ${LIB_PREFIX}${OPENCL_LIBNAME} ${SYCL_LINK_FLAGS})
```
Don't we use DPC++ compiler here? It should have -fsycl by default shouldn't it?
But even if we have to provide explicitly linker option for some reasons I think we need to change it a little bit. I'm not sure I like the idea of providing linker option as one of library because in some cases order of options and libraries is important. I'd suggest at least add -fsycl to the very beginning of the MKL_LINK_SYCL before any libraries.
```suggestion
    set(MKL_LINK_SYCL ${SYCL_LINK_FLAGS} ${MKL_LINK_PREFIX} ${LIB_PREFIX}${MKL_SYCL} ${MKL_LINK_C} ${LIB_PREFIX}${OPENCL_LIBNAME})
```

mmeterel replied:
Thanks for the feedback. We are using clang++ compiler, so it does not have -fsycl by default. For this file, compilation has -fsycl but not link line. 
I will update the sequence with your suggestion.

I will change the order.

mkrainiuk commented on src/blas/backends/mklgpu/CMakeLists.txt:
```diff
@@ -44,7 +45,6 @@ set_target_properties(${LIB_OBJ} PROPERTIES
 )
 target_link_libraries(${LIB_NAME} PUBLIC ${LIB_OBJ})
 
-#Set MKL libraries as not transitive for dynamic
```
I thought it makes sense to comment it here, but if you think it's clear why we are doing next three lines I think it's fine.

mmeterel replied:
Here, I was actually testing something. I will undelete it.

mmeterel commented:
Hi @mkrainiuk  Thanks for the review. I have addressed your feedback. Pls let me know if you have any other reviews.

mmeterel commented:
/intelci: restart

mmeterel commented:
/intelci: run

mmeterel commented:
/intelci: run

mmeterel commented:
/intelci: restart
